src/libsyntax/ext/expand.rs

```diff
@@ -595,16 +675,35 @@ impl<'a, 'b> InvocationCollector<'a, 'b> {
 
     fn collect_attr(&mut self, attr: ast::Attribute, item: Annotatable, kind: ExpansionKind)
                     -> Expansion {
-        self.collect(kind, InvocationKind::Attr { attr: attr, item: item })
+        let invoc_kind = if is_derive_attr(&attr) {
+            InvocationKind::Derive { attr: attr, item: item }
+        } else {
+            InvocationKind::Attr { attr: attr, item: item }
+        };
+
+        self.collect(kind, invoc_kind)
     }
 
     // If `item` is an attr invocation, remove and return the macro attribute.
     fn classify_item<T: HasAttrs>(&mut self, mut item: T) -> (T, Option<ast::Attribute>) {
         let mut attr = None;
+
         item = item.map_attrs(|mut attrs| {
-            attr = self.cx.resolver.find_attr_invoc(&mut attrs);
+            attr = self.cx.resolver.find_attr_invoc(&mut attrs)
+                .or_else(|| {
+                    verify_derive_attrs(self.cx, &mut attrs);
+                    get_legacy_derive(self.cx, &mut attrs)
+                }).or_else(|| {
+                    get_proc_macro_derive(self.cx, &mut attrs)
+                }).or_else(|| {
+                    add_structural_marker(self.cx, &mut attrs);
+                    add_copy_clone_marker(self.cx, &mut attrs);
```

Comment:
One annoying thing is that currently we need some span to add for the new attribute. If we make a `Vec<Name>`, we'll either need to grab the span of the first attribute, or if possible make a dummy span.